doc/designs/quic-design/quic-event-loop.md

```diff
@@ -0,0 +1,160 @@
+# Design Problem: Event Loop
+
+Applications need to be able to react to events on:
+
+* Sockets
+* Connections
+* Streams
+* Timers
+
+Only the OpenSSL library can supply the events for Connections and Streams.
+Events on sockets and timers ultimately come from a system source - the
+application may want to receive the event directly from the system, or from
+OpenSSL.
+
+For the MVP a blocking mode on top of one of the options below is likely to be
+required because that’s the default mode for `s_client`.
+
+## Candidate Solution: Application Selects
```

Comment:
I would say that it is the only viable solution for many applications, in particular the ones which use `boost::asio` or `libuv` for networking. Or, to be more precise, a viable solution for these applications is the solution which decouples the QUIC code from its networking code completely.

That is, an application manages the sockets and socket events completely on its own - including reading and writing data to sockets. Similarly to how TLS is implemented.

That implies that the design should not care about from where data arrived or how data gets sent via network - handling data and sockets is completely application's responsibility. Anything shorter than that would make implementation completely useless for the applications having its own networking middleware.